virtual_ship/sailship.py

```diff
@@ -222,68 +86,348 @@ def sailship(config: VirtualShipConfig):
+) -> None:
+    measurements.drifters.append(
+        Drifter(
+            cruise.spacetime,
+            depth=config.drifter_config.depth,
+            lifetime=config.drifter_config.lifetime,
+        )
+    )
+
 
-    # explitly include final point which is not added by the previous loop
-    sample_points.append(
-        Location(
-            latitude=config.route_coordinates[-1].lat,
-            longitude=config.route_coordinates[-1].lon,
+def _argo_float_task(
+    cruise: _Cruise, measurements: _MeasurementsToSimulate, config: VirtualShipConfig
+) -> None:
+    measurements.argo_floats.append(
+        ArgoFloat(
+            spacetime=cruise.spacetime,
+            min_depth=config.argo_float_config.min_depth,
+            max_depth=config.argo_float_config.max_depth,
+            drift_depth=config.argo_float_config.drift_depth,
+            vertical_speed=config.argo_float_config.vertical_speed,
+            cycle_days=config.argo_float_config.cycle_days,
+            drift_days=config.argo_float_config.drift_days,
         )
     )
 
-    return sample_points
+
+def _verify_waypoints(
+    waypoints: list[Waypoint], ship_speed: float, projection: pyproj.Geod
+) -> None:
+    """
+    Verify waypoints are ordered by time, first waypoint has a start time, and that schedule is feasible in terms of time if no unexpected things happen.
+
+    :param waypoints: The waypoints to check.
+    :param ship_speed: Speed of the ship.
+    :param projection: projection used to sail between waypoints.
+    :raises PlanningError: If waypoints are not feasible or incorrect.
+    """
+    # check first waypoint has a time
+    if waypoints[0].time is None:
+        raise PlanningError("First waypoint must have a specified time.")
+
+    # check waypoint times are in ascending order
+    timed_waypoints = [wp for wp in waypoints if wp.time is not None]
+    if not all(
+        [
+            next.time >= cur.time
+            for cur, next in zip(timed_waypoints, timed_waypoints[1:])
+        ]
+    ):
+        raise PlanningError(
+            "Each waypoint should be timed after all previous waypoints"
+        )
+
+    # check that ship will arrive on time at each waypoint (in case nothing goes wrong)
+    time = waypoints[0].time
+    for wp, wp_next in zip(waypoints, waypoints[1:]):
+        if wp.instrument is InstrumentType.CTD:
+            time += timedelta(minutes=20)
+
+        geodinv: tuple[float, float, float] = projection.inv(
+            wp.location.lon, wp.location.lat, wp_next.location.lon, wp_next.location.lat
+        )
+        distance = geodinv[2]
+
+        time_to_reach = timedelta(seconds=distance / ship_speed)
+        arrival_time = time + time_to_reach
+
+        if wp_next.time is None:
+            time = arrival_time
+        elif arrival_time > wp_next.time:
+            raise PlanningError(
+                f"Waypoint planning is not valid: would arrive too late at a waypoint requiring time {wp_next.time}"
```

Comment:
Can this error also provide _which_ waypoint is invalid? That will help debugging

Reply:
is this better?

```
virtual_ship.planning_error.PlanningError: Waypoint planning is not valid: would arrive too late at a waypoint number 1. location: Location(latitude=-43.081289, longitude=63.743631) time: None instrument: InstrumentType.CTD
```